### ISSUE TITLE
Stuck Key Input Issue and Misdetected keys

### ISSUE THREAD

**Issue by DrJedikiah** [ocornut/imgui#2062]
Latest version of ImGui and every renderer in win32

Lately I made a custom widget which is a hotkey container for assigning keys to jobs. My issue is if you change the active window while you holding a key, key becomes stuck in pressed state and only pressing again that key solves that problem. To demonstrate it, just open the ImGui xxxx example doesn't matter the renderer just win32, and go to the inputs section and keyboard mouse states. After that just try to alt tab (easy demo) or activate another window while you holding a key. You will see that key is flashing like pressing rapidly in "key pressed:" section. That causes to not detecting the proper keystroke.

![screenshot_2](https://user-images.githubusercontent.com/12988451/45063085-ca557480-b0b4-11e8-8208-8b562ca8a7a1.png)

**Comment by ocornut** [comment]
You are right. We don't receive the `WM_KEYUP` events after we lost focus.
Perhaps we could decide on a policy such as clearing the key down state when receiving a `WM_KILLFOCUS` message?

Happy to hear feedback of how people handle this sorts of thing their own engine?

In dear imgui world this becomes a little more involved in the Viewport branch as there are multiple Win32 windows at play. I'll look at it eventually (suggestion/fixes that are compatible with the Viewport branch are welcome).

**Comment by DrJedikiah** [comment]
Thanks for the help, I've tried to clear key down state but it changes to true (the stuck key state) everytime because windows thinks the key is still physically down. So I made something like this and now works perfectly but it's not the most efficient way of doing this I think.

![screenshot_1](https://user-images.githubusercontent.com/12988451/45091958-a9c10500-b11c-11e8-8615-7baecdea9b92.png)

**Comment by DrJedikiah** [comment]
Update:
`case WM_SETFOCUS:
        for ( int i = 0; i < 512; i++ ) {
            UINT virtualKey = (BYTE)i;
            UINT scanCode = MapVirtualKeyA( virtualKey, MAPVK_VK_TO_VSC );
            keybd_event( virtualKey, scanCode, KEYEVENTF_KEYUP, NULL );
        }
        return 0;`

I've refined the code but forgot to publish. Adding this to WndProcHandler will solve the issue.

**Comment by mrdooz** [comment]
Thanks for the easy fix! I was running into the same issue when I was hitting a breakpoint inside my code that responded to InputText returning true, so the debugger stole input and the WM_KEYUP got lost.

**Comment by ocornut** [comment]
#3961 suggested using:
```
    case WM_KILLFOCUS:
        memset(io.KeysDown, 0, sizeof(io.KeysDown));
        return 0;
```
I however mentioned that the problem is that none of those strategy works for multi-viewports, but I'll look for a solution that is multi-viewports compatible.

**Comment by ocornut** [comment]
I pushed the WM_KILLFOCUS suggested fix now : 1491d2c9168a1982ea88dd638509e64ec990daf7

As it happens, in docking + multi-viewports enabled it does make held key flick off for one frame, but this issue has also been present in the GLFW backend and no-one noticed that before, so my conclusion is that it's a much less problematic issue than the one we are fixing now. We can improve on the multi-viewports code later.

**Comment by JoshuaWebb** [comment]
Could/should the same fix be implemented in the backend for `io.MouseDown`? It isn't nearly as common to lose focus while a mouse button is down as it is for the keyboard, so I'm happy to just handle it in my own code, but I ran into it today so thought it was worth asking.
